### PR TITLE
1113.2 - ssp section 1 parsing

### LINKED ISSUE #1113: SSP Active Data (01) File Parsing - As a data analyst, I want my data files to be parsed so that my submission can be validated/accepted by TDP
Description:
Parsing of TANF STT submissions for Stratum Data.

AC: 
- [ ] Data File parsing logic is in place such that data file submission can be completed by user and parsed by system for full submission once errors are resolved.
- [ ] Data Errors tell user specifically where in the file they occur, and therefore what exactly needs to be fixed in order for the submission to be complete
- [ ] TDP gives the user all data errors for a submission in a way they can easily save or export so they can take the issues back to their case management software and fix the issues then recreate an export for resubmission to TDP.
- [ ] Data Errors must be issued in human readable plain language so that an english speaker can easily understand what is needed. 

Tasks: 
- [ ] Sample tasks

DoD: 
 - [ ] minimum 90% code coverage test
 - [ ] 0 linting errors and warnings
 - [ ] 0 accessibility errors
 - [ ] deployed to a dev environment
 - [ ] all significant changes have been documented
 - [ ] free of high and medium security vulnerabilities

Supporting Documentation:
https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_data_report_section1_10_2008.pdf 
SSP-MOE Data Report - Section 1 - Disaggregated Data Collection For Families Receiving Assistance Under The State Separate Programs (Effective 10/1/2008)

- Family Case Data - Revised 10/1/2008
- Adult Data - Revised 10/1/2008
- Child Data - Revised 10/1/2008
https://www.acf.hhs.gov/sites/default/files/documents/ofa/ssp_moe_transmission_format_section_1.pdf

Open Questions:

(any clarifying questions in support of the issue)

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes

Pull request closes #1113
* ssp section 1 parsing, building on M1/M2/M3 models added in #2478 
* rows representing m3 records may actually contain two records, creates a `MultiRecordRowSchema` that allows a row to be parsed into multiple objects.

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
1. Navigate to Data files and upload one of the ssp test files included in this PR
1. Ensure parsing passes and errors logged in the console match what is expected.

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] [**_insert ACs here_**]
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [x] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+              required=True, validators=[]),
+        Field(name='FAMILY_CAP', type='number', startIndex=109, endIndex=110, required=True, validators=[
+        ]),
+        Field(name='REDUCTIONS_ON_RECEIPTS', type='number', startIndex=110, endIndex=111,
+              required=True, validators=[]),
+        Field(name='OTHER_NON_SANCTION', type='number', startIndex=111, endIndex=112,
+              required=True, validators=[]),
+        Field(name='WAIVER_EVAL_CONTROL_GRPS', type='number', startIndex=112, endIndex=113,
+              required=True, validators=[]),
+        # Field(name='FAMILY_EXEMPT_TIME_LIMITS', type='number', startIndex=114, endIndex=116,
```
Comment (elipe17) [inline tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py]: Can we remove the commented fields?

**Review by elipe17 — APPROVED** [review]
Everything looks good to me. Just one cleanliness question. Also, it is just worth noting if you try to upload the `ssp_section1_datafile.txt` file you get a 413 code. I'm pretty sure Mo has fixed this in another PR but just wanted to bring it to the surface.

File: tdrs-backend/tdpservice/parsers/parse.py [hunk 2]
```diff
@@ -75,21 +85,48 @@ def parse_datafile(datafile):
...
 
         schema = get_schema(line, section, schema_options)
-        record_is_valid, record_errors = parse_datafile_line(line, schema)
 
-        if not record_is_valid:
-            errors[line_number] = record_errors
+        if isinstance(schema, util.MultiRecordRowSchema):
+            records = parse_multi_record_line(line, schema)
+
+            n = 0
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/parse.py]: Since this is for multi-record lines, n doesn't correlate with line number but rather interline record count, correct? Could we alter the name for clarity?

Alternatively, do we even need a named entry? What's the advantage/is this used? I believe we keep multiple errors for ordinary lines, why not just keep that same structure and get the info needed from ParserError which has name, startIndex, case number?

File: tdrs-backend/tdpservice/parsers/parse.py [hunk 3]
```diff
@@ -75,21 +85,48 @@ def parse_datafile(datafile):
...
 
+def parse_multi_record_line(line, schema):
+    """Parse and validate a datafile line using MultiRecordRowSchema."""
+    if schema:
+        records = schema.parse_and_validate(line)
+
+        for r in records:
+            record, record_is_valid, record_errors = r
+
+            if record:
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/parse.py:116]: If record is None here, we loop to next r? Did this come up in testing?

File: tdrs-backend/tdpservice/parsers/schema_defs/ssp/m3.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    postparsing_validators=[],
+    fields=[
+        Field(name='RecordType', type='string', startIndex=0, endIndex=2,
+              required=True, validators=[]),
+        Field(name='RPT_MONTH_YEAR', type='number', startIndex=2, endIndex=8,
+              required=True, validators=[]),
+        Field(name='CASE_NUMBER', type='string', startIndex=8, endIndex=19,
+              required=True, validators=[]),
+        # Field(name='FIPS_CODE', type='string', startIndex=8, endIndex=19, required=True, validators=[
+        # ]),
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/schema_defs/ssp/m3.py]: Why is this commented out? I note it contains same indices as CASE_NUMBER.

File: tdrs-backend/tdpservice/parsers/util.py [hunk 5]
```diff
@@ -116,11 +143,11 @@ def run_field_validators(self, instance):
...
                 value = None
                 if isinstance(instance, dict):
-                    value = instance[field.name]
+                    value = instance.get(field.name, None)
                 else:
-                    value = getattr(instance, field.name)
+                    value = getattr(instance, field.name, None)
 
-                if field.required and value != (' '*(field.endIndex-field.startIndex)):
+                if field.required and not value_is_empty(value, field.endIndex-field.startIndex):
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/util.py:150]: Love the readability improvement here!

File: tdrs-backend/tdpservice/parsers/schema_defs/ssp/m3.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+
+from ...util import MultiRecordRowSchema, RowSchema, Field
+from ... import validators
+from tdpservice.search_indexes.models.ssp import SSP_M3
+
+first_part_schema = RowSchema(
+    model=SSP_M3,
+    preparsing_validators=[
+        # validators.hasLength(150),  # unreliable.
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/schema_defs/ssp/m3.py]: As these can be multi-record, this line should be removed given the business case, yes?

File: tdrs-backend/tdpservice/parsers/parse.py (hunk 3 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/parse.py:116]: this would come up any time a preparsing error is encountered for a line, weather that's in a `MultiRecordRowSchema` or a regular `RowSchema`. the test `test_parse_big_file` covers this case in the expected number of errors assertion.

File: tdrs-backend/tdpservice/parsers/schema_defs/ssp/m3.py (hunk 4 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/schema_defs/ssp/m3.py]: it was leftover from a bad copy/paste job :) i've removed it

File: tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        Field(name='FAILURE_TO_COMPLY', type='number', startIndex=99, endIndex=100,
+              required=True, validators=[]),
+        Field(name='OTHER_SANCTION', type='number', startIndex=100, endIndex=101,
+              required=True, validators=[]),
+        Field(name='RECOUPMENT_PRIOR_OVRPMT', type='number', startIndex=101, endIndex=105,
+              required=True, validators=[]),
+        Field(name='OTHER_TOTAL_REDUCTIONS', type='number', startIndex=105, endIndex=109,
+              required=True, validators=[]),
+        Field(name='FAMILY_CAP', type='number', startIndex=109, endIndex=110, required=True, validators=[
+        ]),
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py]: newline here because of line-length lint rule?

**Comment by andrew-jameson** [comment]
Uploading test datafile 'ADS.E2J.FTP1.TS06' results in consistent 'no schema selected' errors:

```
   2023-05-04T12:14:50.47-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:14:50,475 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM1.TS01
   2023-05-04T12:14:50.47-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:14:50,475: INFO/ForkPoolWorker-1] DataFile parsing started for file ADS.E2J.NDM1.TS01
   2023-05-04T12:14:50.59-0400 [APP/PROC/WEB/0] OUT 10.255.175.145 - - [04/May/2023:16:14:50 +0000] "POST /v1/data_files/ HTTP/1.0" 201 283 "https://tdp-frontend-raft.app.cloud.gov/data-files" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/112.0"
   2023-05-04T12:14:50.73-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:14:50 +0000] [574] [DEBUG] POST /v1/logs/
   2023-05-04T12:14:50.75-0400 [APP/PROC/WEB/0] OUT 10.255.175.145 - - [04/May/2023:16:14:50 +0000] "POST /v1/logs/ HTTP/1.0" 200 9 "https://tdp-frontend-raft.app.cloud.gov/data-files" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/112.0"
   2023-05-04T12:15:16.24-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:15:16 +0000] [574] [DEBUG] Closing connection.
   2023-05-04T12:15:19.82-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:15:19,827 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 1828 errors: {3: ['No schema selected.'], 4: ['No schema selected.'], 6: ['No schema selected.'], .... ad naseum for every line until trailer.
   ```

File: tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py (hunk 7 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py]: missed it during formatting. fixed!

File: tdrs-backend/tdpservice/parsers/parse.py (hunk 2 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/parse.py]: the goal was just to add some structure/clarity to the returned error object for my own testing. you're right in that we don't really use the returned object, and flattening it would still contain all the relevant data. if you'd like to discuss the structure of the returned errors object, i'm open to having a broader conversation about the entire thing.

**Comment by andrew-jameson** [comment]
> Uploading test datafile 'ADS.E2J.FTP1.TS06' results in consistent 'no schema selected' errors:
> 
> ```
>    2023-05-04T12:14:50.47-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:14:50,475 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM1.TS01
>    2023-05-04T12:14:50.47-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:14:50,475: INFO/ForkPoolWorker-1] DataFile parsing started for file ADS.E2J.NDM1.TS01
>    2023-05-04T12:14:50.59-0400 [APP/PROC/WEB/0] OUT 10.255.175.145 - - [04/May/2023:16:14:50 +0000] "POST /v1/data_files/ HTTP/1.0" 201 283 "https://tdp-frontend-raft.app.cloud.gov/data-files" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/112.0"
>    2023-05-04T12:14:50.73-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:14:50 +0000] [574] [DEBUG] POST /v1/logs/
>    2023-05-04T12:14:50.75-0400 [APP/PROC/WEB/0] OUT 10.255.175.145 - - [04/May/2023:16:14:50 +0000] "POST /v1/logs/ HTTP/1.0" 200 9 "https://tdp-frontend-raft.app.cloud.gov/data-files" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/112.0"
>    2023-05-04T12:15:16.24-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:15:16 +0000] [574] [DEBUG] Closing connection.
>    2023-05-04T12:15:19.82-0400 [APP/PROC/WEB/0] ERR [2023-05-04 16:15:19,827 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 1828 errors: {3: ['No schema selected.'], 4: ['No schema selected.'], 6: ['No schema selected.'], .... ad naseum for every line until trailer.
> ```

Discussed offline. Errors are due to not being implemented yet. Could maybe change the errors to be more direct but I don't feel strongly about it.

**Review by andrew-jameson — APPROVED** [review]

**Review by George-Hudson — APPROVED** [review]

**Review by ADPennington — APPROVED** [review]
@jtimpe this looks good! Please note the following about testing file uploads (discussed with @andrew-jameson and @raftmsohani):

- i'm making it a point to test parsing PRs with large files, which are saved in [raft's sharepoint site](https://gorafttech.sharepoint.com/sites/TDRSResearchDesign/Shared%20Documents/Forms/AllItems.aspx?id=%2Fsites%2FTDRSResearchDesign%2FShared%20Documents%2FDev%2Fdev%2FParsing%2Fdata&viewid=981949f0%2D98e1%2D4ace%2D8da8%2D436aca592a46)
- the test results were good with legitimate SSP file submissions. 
- the test results were more inconsistent when submitting large TANF files in the SSP file slots. I know that this is an edge case, but we should keep an eye on what happens when someone submits a file in the wrong spot. I know some of this will be handled in error/validation tickets as well as tickets like #2346 and #2282 

----

example of failure (submitting large tanf files in SSP file slots):
1113part2

backend log:

```
12:39:41.524: [APP/PROC/WEB.0][2023-05-16 16:39:41,524: ERROR/MainProcess] Task handler raised error: WorkerLostError('Worker exited prematurely: signal 9 (SIGKILL) Job: 87.')
12:39:41.524: [APP/PROC/WEB.0]Traceback (most recent call last):
12:39:41.524: [APP/PROC/WEB.0] File "/home/vcap/deps/1/python/lib/python3.10/site-packages/billiard/pool.py", line 1265, in mark_as_worker_lost
12:39:41.524: [APP/PROC/WEB.0] raise WorkerLostError(
12:39:41.524: [APP/PROC/WEB.0]billiard.exceptions.WorkerLostError: Worker exited prematurely: signal 9 (SIGKILL) Job: 87.
```

frontend log:

```
12:41:30.179: [APP/PROC/WEB.0]2023/05/16 16:41:30 [error] 115#0: *11591 upstream prematurely closed connection while reading response header from upstream, client: 2601:140:8c00:55b:5920:aab7:91bc:67c6, server: , request: "POST /v1/data_files/ HTTP/1.1", upstream: "http://10.255.242.183:8080/v1/data_files/", host: "tdp-frontend-qasp.app.cloud.gov", referrer: "https://tdp-frontend-qasp.app.cloud.gov/data-files"
12:41:30.439: [APP/PROC/WEB.0]2023/05/16 16:41:30 [error] 115#0: *11591 open() "/home/vcap/app/public/50x.html" failed (2: No such file or directory), client: 2601:140:8c00:55b:5920:aab7:91bc:67c6, server: , request: "POST /v1/data_files/ HTTP/1.1", upstream: "http://10.255.242.183:8080/v1/data_files/", host: "tdp-frontend-qasp.app.cloud.gov", referrer: "https://tdp-frontend-qasp.app.cloud.gov/data-files"
```
